akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -627,6 +627,75 @@ final class Broadcast[T](val outputPorts: Int, val eagerCancel: Boolean) extends
 
 }
 
+object WireTap {
+  private val singleton = new WireTap[Nothing]
+  def apply[T]() = singleton.asInstanceOf[WireTap[T]]
+}
+
+/**
+ * Fan-out the stream to two output streams - a 'main' and a 'tap' one - emitting each incoming upstream element
+ * to all downstream consumers.
+ *
+ * '''Emits when''' element is available and demand exists from the 'main' output; the element will
+ * also be sent to the 'tap' output if there is demand.
+ *
+ * '''Backpressures when''' the 'main' output backpressures
+ *
+ * '''Completes when''' upstream completes
+ *
+ * '''Cancels when''' the 'main' output cancels
+ *
+ */
+private[stream] final class WireTap[T] extends GraphStage[FanOutShape2[T, T, T]] {
+  val in: Inlet[T] = Inlet[T]("WireTap.in")
+  val outMain: Outlet[T] = Outlet[T]("WireTap.outMain")
+  val outTap: Outlet[T] = Outlet[T]("WireTap.outTap")
+  override def initialAttributes = DefaultAttributes.broadcast
+  override val shape: FanOutShape2[T, T, T] = new FanOutShape2(in, outMain, outTap)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    private var pendingTap: Option[T] = None
```

Comment:
Use `akka.util.OptionVal` here instead, it basically hides `null` and avoids the `Some` allocation for each pull.

Reply:
That would, however, require changing the type of this to `[T :> Null]`; i.e., `AnyVal`s would no longer be allowed. That seems undesirable.

Reply:
So, reverted to using a regular Option, unless you know of some way to get around this issue? 🙂